### ISSUE TITLE
Allow YouTube sync to delete older videos when out of space

### ISSUE THREAD

**Issue by TILvids** [Chocobozzz/PeerTube#5642]
### Describe the problem to be solved

As a user, I want to synchronize my YouTube channel to a PeerTube instance. However, my PeerTube instance keeps running out of space.

### Describe the solution you would like

In a perfect world I would have unlimited space on my instance like YouTube, but that is not really practical. The next best solution would be for my PeerTube instance to delete the oldest video to make room for the newest when I'm out of space (or make this a user-toggleable option).

**Comment by test2a** [comment]
I dont know.....
Maybe it should be responsibility of the user or admin to delete old videos. 
Do you want it to be like the CCTV style "keep only latest" so it was will download 10, run out of space, delete first, download 11th, run out of space, delete 2nd, on and on?

Do we have an order of sync so download oldest first or latest first or most popular first because that is more relevant for sync when storage Is consideration.

**Comment by TILvids** [comment]
So I think it should probably be a channel-admin option for them to toggle. The reason why I think this is useful is that, at least for now, PeerTube is often not the primary focus of creators; at best, they're just trying to support the ecosystem, but their primary focus remains their YouTube channel. That being the case, they don't want to have to spend time triaging their PeerTube channel because running their primary YouTube channel takes a lot of effort.

At the same time, most PeerTube instances are limited in space. Users can easily upload TB's of videos on their YouTube channel, whereas a that might be a PeerTube instance's entire storage capacity. When the user runs out of storage space on their channel, they will most likely just abandon their PeerTube channel, and no new content will ever appear.

So that being the case, it'd be better for their newer content to keep showing as they upload it, as opposed to their channel just freezing in time when space is out. I'd even propose that this option should be enabled by default, with the channel-admin having the option to turn it off.

**Comment by emansom** [comment]
Deleting all other resolutions except the highest quality/original one could also be an option to reduce space.

Automating that would be less destructive than deleting randomly based on LRU.

**Comment by mtwb47** [comment]
An option for autio delete would be, I think, the best way to go. That way if the creator wants to do it manually, they can, but if not, they can set it up and forget about it. 

That last bit may be a reason *not* to do it, as you don't really *want* creators to set it and forget it. Then again, the less friction you have in the way of YouTubers using Peertube, the more you'll get using it.

**Comment by TILvids** [comment]
If you want to simplify the logic behind it, I would pick simply oldest out, newest in. That keeps the content fresh. Most of these YouTubers aren't relying on the PeerTube instance to back up their content, and for most instances that's simply an unrealistic expectation anyway.

If you don't mind making the user-experience a bit more complicated, you could allow the channel admin to choose which of these scenarios they want. But for me, I think simplifying it to "oldest out, newest in" is probably good enough to cover most scenarios.

**Comment by maltfield** [comment]
This would be a great feature, and I think the admin should be able to choose from several algorithms to decide which videos to delete:

1. Delete Oldest Videos First
2. Delete Least Viewed First
2a. In the past 7 days
2b. In the past 30 days
2c. In the past year
3. Delete Least "Liked" (in past 7/30/365 days)
4. Delete Largest Videos First
5. Delete Random Video

**Comment by ROBERT-MCDOWELL** [comment]
7. delete from a specific channel with start / end date too

**Comment by RokeJulianLockhart** [comment]
> 2. Delete Least Viewed First
>    1. In the past 7 days
>    1. In the past 30 days
>    1. In the past year
> 3. Delete Least "Liked" (in past 7/30/365 days)

@maltfield, instead of hardcoding arbitrary amounts of specific measures, just provide a free-form integer entry form with a choice of measure (day and year, etcetera). Likewise, for the undermentioned:

> 4. Delete Largest Videos (in bytes) First
> 5. Delete Longest Videos (in seconds) First

**Comment by maltfield** [comment]
@RokeJulianLockhart agreed. But then it should probably be in some very small unit (eg seconds, bytes) or the unit should be a dropdown menu (eg days, months, megabytes, gigabytes)

**Comment by RokeJulianLockhart** [comment]
@maltfield, I'd say the latter:

> the unit should be a dropdown menu (eg days, months, megabytes, gigabytes)

It generally ensures that the implementation is extensible.
